flow/layers/cacheable_layer.cc

```diff
@@ -0,0 +1,46 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/flow/layers/cacheable_layer.h"
+#include "flutter/flow/layers/display_list_layer.h"
+#include "flutter/flow/layers/layer.h"
+#include "flutter/flow/layers/picture_layer.h"
+#include "flutter/flow/raster_cacheable_entry.h"
+namespace flutter {
+
+Cacheable::AutoCache Cacheable::AutoCache::Create(Cacheable* cacheable,
+                                                  PrerollContext* context,
+                                                  const SkMatrix& matrix) {
+  auto* entry = context->raster_cached_entries
+                    .emplace_back(RasterCacheableEntry::MarkLayerCacheable(
+                        cacheable, *context, matrix))
+                    .get();
+  return AutoCache(cacheable, entry, context, matrix);
+}
+
+Cacheable::AutoCache Cacheable::AutoCache::Create(
+    DisplayListLayer* display_list_layer,
+    PrerollContext* context,
+    const SkMatrix& matrix,
+    SkPoint offset) {
+  auto* entry =
+      context->raster_cached_entries
+          .emplace_back(RasterCacheableEntry::MarkDisplayListCacheable(
+              display_list_layer->display_list(), *context, matrix, offset))
+          .get();
+  return AutoCache(display_list_layer, entry, context, matrix);
+}
+
+Cacheable::AutoCache Cacheable::AutoCache::Create(PictureLayer* picture_layer,
+                                                  PrerollContext* context,
+                                                  const SkMatrix& matrix,
+                                                  SkPoint offset) {
+  auto* entry = context->raster_cached_entries
+                    .emplace_back(RasterCacheableEntry::MarkSkPictureCacheable(
+                        picture_layer->picture(), *context, matrix, offset))
+                    .get();
+  return AutoCache(picture_layer, entry, context, matrix);
```

Comment:
Is raster_cached_entries available if there is no raster_cache in the context? Should it be? It's not useful if there is no raster cache. Note that we don't need any touching or prepare calls if the raster cache is missing as it usually means this is an out-of-band request for a layer tree to paint (for example, a `toImage()` call on the layer tree) so it should just do the painting regularly.

These Create methods should check various conditions that preclude caching, such as "raster_cache entry is null" which may also mean "raster_cache_entries is null". If so, then they should not do the emplace_back and return an AutoCache that is benign (for instance, it could have a null entry and/or cacheable and then the destructor just wouldn't bother doing any work).